debian/wviewconfig.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH WVIEWCONFIG 1 "November 19, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
wviewconfig \- script to configure a wview installation
.SH SYNOPSIS
.B wviewconfig
.RI
.br
.SH DESCRIPTION
This manual page describes the use of
.B wviewconfig
.
Configure a wview installation.
.BR
.BR Note: wview should not be running when running this script.
.BR Note: must be executed on the source wview server.
.SH OPTIONS
There are no options.
.TP
.SH SEE ALSO
.BR wviewhtmlconfig (1),
.BR /usr/share/doc/wview/wview-User-Manual.html
.br
.SH AUTHOR
wviewconfig was written by Mark Teel.
.PP
This manual page was written by Mark Teel <mteel2005@gmail.com>,
for the Debian project (and may be used by others).


ChangeLog
RELEASE VERSION HISTORY (newest to oldest)
------------------------------------------


5.21.5    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Remove debug message.




5.21.4    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Added stale pid file cleanup to the APT package start script.

2)  WMRUSB HID interface changes. Changed reader thread to blocking IO.
    Mitigated the console occasionally blasting out bogus data. Now zero
    packet loss. Fixed issue with pressure sanity check. Added sanity
    check for bogus rain accumulator values from the WMR console.




5.21.3    03-31-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Further WMRUSB interface improvements. Better packet framing.




5.21.2    03-30-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Updated debian APT postinst script to include copying the exfoliation skin
    for new installs.

2)  Enabled checksum processing for WMR stations.

3)  Changed LOOP windspeed items to floats for better precision.
    NOTE: If you use custom datafeed clients the LOOP data structure has changed.

4)  Added temp only packet support for WMRUSB stations. This adds support for
    add-on sensors such as the pool sensor.

5)  Fixed APT installs changing the wview.log file ownership incorrectly. Also
    restarts rsyslog automatically if a new install.




5.21.1    03-13-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Modified HILOW database storage so higher precision is used for cumulative
    data such as rain and ET.

2)  Fixed issue with yearly historical charts when generated at wview start
    (as opposed to at midnight). The current day was erroneously being included in
    the history data. This was most apparent with the UV and Solar Radiation
    charts.

3)  Fixed bug introduced in 5.21.0 release causing NULL wind direction to have
    360 added to it (to normalize calibration result). A value of -99640 was
    being stored/displayed.




5.21.0    03-05-2014
--------------------

radlib version required: 2.12.0 or newer

1)  WMRUSB improvements. Further reverse engineering of the undocumented protocol. Reset and
    heartbeat messages now working. Packet reception more stable. Initial packet acquisition
    faster resulting in faster startup.

2)  Modify CWOP packet header for new stricter compliance requirements of CWOP servers.

3)  Fix log message time_t bug for 64-bit platforms.

4)  Add timeout parameter for ssh rules. Added rows to config database.

5)  Updated data feed conversion logic to handle ARCHIVE_VALUE_NULL in loop data.

6)  Added flat line sensor value detection for outside temp, wind direction and wind speed.
    If there are no changes over a 4 hour period an email alert will be sent.

7)  Added Station Name, City and State to the wviewmgmt System Status page header.

8)  Added logic to SSH timer to fixup next timer interval if system time has skewed or
    if the ssh transfer took longer than 2 minutes.

9)  Fixed TE923 rain calculation to assume each tip is 1/36" instead of 1mm.

10) Fixed a bug for Vantage Pro stations which could cause the last archive interval of
    the previous day to be included in the current day HILOW data. This occurred if the
    midnight archive record was provided late by the console.

11) Added wet bulb temperature calculation and html tag --WetBulbTemp--. Added to all
    "readings.incx" templates.

12) Added email alert notification if no loop data has been received for a given archive
    interval. This applies to all stations except the Davis stations (alerts already
    exist for them).

13) Updated start scripts to detect and delete stale pid files.

14) Modified WH1080 data parsing to assume a humidity of 9% if the value from the
    station is undefined (0xff).




5.20.2    03-21-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added back "dailyrainin" for http reporting. They do no computation based on hourly rain values.

2)  Fixed a typo in the wviewmgmt station.php file. 




5.20.1    03-20-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Modified http rain submission (Wunderground and Weatherforyou) so only the
    "rainin" parameter is sent so daily rain will be correct. 




5.20.0    03-17-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added "wait_for_time_set" function to most wview start script examples so 
    wview daemons are not started until the system time is set. This applies to
    platforms without a real time clock which must reset their time during 
    startup.

2)  Updated the User Manual section on skin selection to better describe enabling
    extended data generation (UV, ET, radiation).

3)  Updated the debian post install script to try harder to find the default http
    document root location for creation of symbolic links to wviewmgmt and the
    weather home page.

4)  Added new configuration option to optionally disable using the Vantage Pro 
    console archive records and instead autogenerate them from the LOOP data.
    This is how all other stations work. It also supports non-standard console
    interfaces (those not using the datalogger on the console).

5)  Updated the Weatherforyou URL for data submissions.

6)  Fixed bug with DataFeed clients that caused an infinite loop if you
    restarted wview while there was a datafeed client connection.

7)  Fixed bug with Alarm Type configuration introduced in 5.19.0.

8)  Changed TWI station to compute rain fall rate as it is reported to not
    be provided by the station.

9)  Added socket retry logic to the station ethernet interface (used by the
    Vantage Pro IP datalogger and any serial-to-ethernet device).

10) Modified CWOP host name to IP address resolution so that 'getaddrinfo' is
    used instead of 'gethostbyname' and all results are tried in the order they 

wviewconfig/wviewconfig.sh
################################################################################
#
# File:           wviewconfig.sh
#
# Description:    Provide a script to interactively configure a wview 
#                 installation.
#
# Usage:          (must be root)
#                 wviewconfig
#                 wviewconfig get
#                 wviewconfig set [new_config_path]
#
# History:
# Engineer	  Date	    Ver   Comments
# MS Teel	  06/20/05   1    Original
# J Barber	  02/24/06   2    Partitioned into functions; added get/set
# MS Teel	  02/25/06   3    Tweaked arg and function names
# MS Teel	  06/21/08   4    Better station type support
# MS Teel	  08/23/08   5    Modify to use sqlite config database
#
################################################################################


################################################################################
#################################  M A C R O S  ################################
################################################################################
WVIEWD_PID=$WVIEW_DATA_DIR/wviewd.pid
CFG_STATION_TYPE=$STATION_TYPE

STATION_BIN=`cat $WVIEW_CONF_DIR/wview-binary`
if [ "$STATION_BIN" = "wviewd_vpro" ]; then
    STATION_TYPE="VantagePro"
else
    if [ "$STATION_BIN" = "wviewd_wxt510" ]; then
        STATION_TYPE="WXT510"
    else
        if [ "$STATION_BIN" = "wviewd_twi" ]; then
            STATION_TYPE="TWI"
        else
            if [ "$STATION_BIN" = "wviewd_ws2300" ]; then
                STATION_TYPE="WS-2300"
            else
                if [ "$STATION_BIN" = "wviewd_wmr918" ]; then
                    STATION_TYPE="WMR918"
                else
                    if [ "$STATION_BIN" = "wviewd_wmrusb" ]; then
                        STATION_TYPE="WMRUSB"
                    else
                        if [ "$STATION_BIN" = "wviewd_wh1080" ]; then
                            STATION_TYPE="WH1080"
                        else
                            if [ "$STATION_BIN" = "wviewd_te923" ]; then
                                STATION_TYPE="TE923"
                            else
                                if [ "$STATION_BIN" = "wviewd_virtual" ]; then
                                    STATION_TYPE="Virtual"
                                else
                                    STATION_TYPE="Simulator"
                                fi
                            fi
                        fi
                    fi
                fi
            fi
        fi
    fi
fi

################################################################################
#######################  D E F I N E  F U N C T I O N S  #######################
################################################################################

show_usage()
{
    echo ""
    echo "wviewconfig"
    echo "    Configures wview interactively"
    echo "wviewconfig get"
    echo "    Prints current settings in text format to stdout"
    echo "wviewconfig set [new_config_path]"
    echo "    Takes text format file of configuration at [new_config_path] and applies"
    echo "    it to the wview configuration database - it can be a partial or full list"
    echo ""
}


print_config()
{
	if [ -f $WVIEW_CONF_DIR/wview-conf.sdb ]; then
		# get settings from the sqlite database:
		echo ".separator =" > $WVIEW_CONF_DIR/commands.sql
		echo "select name,value from config;" >> $WVIEW_CONF_DIR/commands.sql
		echo ".read $WVIEW_CONF_DIR/commands.sql" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb
		rm -rf $WVIEW_CONF_DIR/commands.sql
	else
		echo "wview Configuration database $WVIEW_CONF_DIR/wview-conf.sdb NOT FOUND - ABORTING!"
		exit 1
	fi
}


interactive_intro()
{
    echo "################################################################################"
    echo " !!!!!!!!!!!!!!!!         READ THIS BEFORE PROCEEDING         !!!!!!!!!!!!!!!!"
    echo ""
    echo "--> System Configuration for wview"
    echo ""
    echo "--> Values in parenthesis are your existing values (if they exist) or defaults - "
    echo "    they will be used if you just hit enter at the prompt..."
    echo ""
    echo "--> Note: This script will save the existing wview-conf.sdb file to" 
    echo "          $WVIEW_CONF_DIR/wview-conf.old before writing the new file" 
    echo "          based on your answers here - if that is not what you want, hit CTRL-C now to"
    echo "          abort this script!"
    echo ""
    echo "################################################################################"
    echo ""
    echo -n "pausing 3 seconds "
    sleep 1
    echo -n "."
    sleep 1
    echo -n "."
    sleep 1
    echo "."
    echo ""
}


get_wview_conf_interactive()
{
	if [ -f $WVIEW_CONF_DIR/wview-conf.sdb ]; then
		# get settings from the sqlite database:
		echo ".separator |" > $WVIEW_CONF_DIR/commands.sql
		echo "select name,value,description,dependsOn from config;" >> $WVIEW_CONF_DIR/commands.sql
		echo ".read $WVIEW_CONF_DIR/commands.sql" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb > $WVIEW_CONF_DIR/parms.out
		rm -rf $WVIEW_CONF_DIR/commands.sql
	
		# Construct the editor script:
		echo "#!/bin/sh"											> $WVIEW_CONF_DIR/editparm
		echo "if [ \"\" != \"\$4\" ]; then"									>> $WVIEW_CONF_DIR/editparm
		echo "    echo \"select value from config where name='\$4';\" > $WVIEW_CONF_DIR/cmnd.sql"		>> $WVIEW_CONF_DIR/editparm
		echo "    echo \".read $WVIEW_CONF_DIR/cmnd.sql\" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb > $WVIEW_CONF_DIR/value.out"	>> $WVIEW_CONF_DIR/editparm
		echo "    IS_ENABLED=\`cat $WVIEW_CONF_DIR/value.out\`"							>> $WVIEW_CONF_DIR/editparm
		echo "    rm -rf $WVIEW_CONF_DIR/cmnd.sql $WVIEW_CONF_DIR/value.out"					>> $WVIEW_CONF_DIR/editparm
		echo "else"												>> $WVIEW_CONF_DIR/editparm
		echo "    IS_ENABLED=yes"										>> $WVIEW_CONF_DIR/editparm
		echo "fi"												>> $WVIEW_CONF_DIR/editparm
		echo "if [ \"\$IS_ENABLED\" != \"yes\" ]; then"								>> $WVIEW_CONF_DIR/editparm
		echo "    exit 0"											>> $WVIEW_CONF_DIR/editparm
		echo "fi"												>> $WVIEW_CONF_DIR/editparm
		echo "NEWVAL=\$2"											>> $WVIEW_CONF_DIR/editparm
		echo "echo \"-------------------------------------------------------------\""				>> $WVIEW_CONF_DIR/editparm
		echo "echo \"\$3\""											>> $WVIEW_CONF_DIR/editparm
		echo "echo \"PARAMETER: \$1\""										>> $WVIEW_CONF_DIR/editparm
		echo "echo -n \"(\$2): \""										>> $WVIEW_CONF_DIR/editparm
		echo "read INVAL"											>> $WVIEW_CONF_DIR/editparm
		echo "if [ \"\" != \"\$INVAL\" ]; then"									>> $WVIEW_CONF_DIR/editparm
		echo "    echo \"update config set value='\$INVAL' where name='\$1';\" > $WVIEW_CONF_DIR/cmnd.sql"	>> $WVIEW_CONF_DIR/editparm
		echo "    echo \".read $WVIEW_CONF_DIR/cmnd.sql\" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb"		>> $WVIEW_CONF_DIR/editparm
		echo "    rm -rf $WVIEW_CONF_DIR/cmnd.sql"								>> $WVIEW_CONF_DIR/editparm
		echo "    if [ \"\$1\" = \"STATION_TYPE\" ]; then"								>> $WVIEW_CONF_DIR/editparm
		echo "        if [ \"\$INVAL\" = \"VantagePro\" ]; then"						>> $WVIEW_CONF_DIR/editparm
		echo "            echo \"wviewd_vpro\" > $WVIEW_CONF_DIR/wview-binary"			>> $WVIEW_CONF_DIR/editparm
		echo "        else"			>> $WVIEW_CONF_DIR/editparm
		echo "            if [ \"\$INVAL\" = \"WXT510\" ]; then"						>> $WVIEW_CONF_DIR/editparm
		echo "                echo \"wviewd_wxt510\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "            else"			>> $WVIEW_CONF_DIR/editparm
		echo "                if [ \"\$INVAL\" = \"WS-2300\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                    echo \"wviewd_ws2300\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                else"			>> $WVIEW_CONF_DIR/editparm
		echo "                    if [ \"\$INVAL\" = \"WMR918\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                        echo \"wviewd_wmr918\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                    else"			>> $WVIEW_CONF_DIR/editparm
		echo "                        if [ \"\$INVAL\" = \"WMRUSB\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                            echo \"wviewd_wmrusb\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                        else"			>> $WVIEW_CONF_DIR/editparm
		echo "                            if [ \"\$INVAL\" = \"WH1080\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                                echo \"wviewd_wh1080\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                            else"			>> $WVIEW_CONF_DIR/editparm
		echo "                                if [ \"\$INVAL\" = \"Simulator\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                                    echo \"wviewd_sim\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                                else"			>> $WVIEW_CONF_DIR/editparm
		echo "                                    if [ \"\$INVAL\" = \"TWI\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                                        echo \"wviewd_twi\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                                    else"			>> $WVIEW_CONF_DIR/editparm
		echo "                                        if [ \"\$INVAL\" = \"TE923\" ]; then"					>> $WVIEW_CONF_DIR/editparm
		echo "                                            echo \"wviewd_te923\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                                        else"			>> $WVIEW_CONF_DIR/editparm
		echo "                                            if [ \"\$INVAL\" = \"Virtual\" ]; then"				>> $WVIEW_CONF_DIR/editparm
		echo "                                                echo \"wviewd_virtual\" > $WVIEW_CONF_DIR/wview-binary"	>> $WVIEW_CONF_DIR/editparm
		echo "                                            fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                                        fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                                    fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                                fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                            fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                        fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                    fi"												>> $WVIEW_CONF_DIR/editparm
		echo "                fi"												>> $WVIEW_CONF_DIR/editparm
		echo "            fi"												>> $WVIEW_CONF_DIR/editparm
		echo "        fi"												>> $WVIEW_CONF_DIR/editparm
		echo "    fi"												>> $WVIEW_CONF_DIR/editparm
		echo "fi"												>> $WVIEW_CONF_DIR/editparm
		chmod +x $WVIEW_CONF_DIR/editparm

        cd $WVIEW_CONF_DIR

		# Edit parms one at a time:
		gawk -F"|" '{ 
		sysstring=sprintf("./editparm \"%s\" \"%s\" \"%s\" \"%s\"", $1, $2, $3, $4)
		system(sysstring)
		}' $WVIEW_CONF_DIR/parms.out
	
		rm -rf $WVIEW_CONF_DIR/editparm $WVIEW_CONF_DIR/parms.out
	else
		echo "wview Configuration database $WVIEW_CONF_DIR/wview-conf.sdb NOT FOUND - ABORTING!"
		exit 1
	fi
}


set_config_from_file()
{
	if [ -f $WVIEW_CONF_DIR/wview-conf.sdb ]; then
		# Construct the update script:
		echo "#!/bin/sh"											> $WVIEW_CONF_DIR/updateparm
		echo "if [ \"\" != \"\$2\" ]; then"									>> $WVIEW_CONF_DIR/updateparm
		echo "    echo \"update config set value='\$2' where name='\$1';\" > $WVIEW_CONF_DIR/cmnd.sql"		>> $WVIEW_CONF_DIR/updateparm
		echo "    echo \".read $WVIEW_CONF_DIR/cmnd.sql\" | sqlite3 $WVIEW_CONF_DIR/wview-conf.sdb"		>> $WVIEW_CONF_DIR/updateparm
		echo "    rm -rf $WVIEW_CONF_DIR/cmnd.sql"								>> $WVIEW_CONF_DIR/updateparm
		echo "fi"												>> $WVIEW_CONF_DIR/updateparm
		chmod +x $WVIEW_CONF_DIR/updateparm

        cd $WVIEW_CONF_DIR

		# Update parms one at a time:
		gawk -F"=" '{ 
		sysstring=sprintf("./updateparm \"%s\" \"%s\"", $1, $2)
		system(sysstring)
		}' $SET_FILE
	
		rm -rf $WVIEW_CONF_DIR/updateparm
	else
		echo "wview Configuration database $WVIEW_CONF_DIR/wview-conf.sdb NOT FOUND - ABORTING!"
		exit 1
	fi	
}


################################################################################
##################  S C R I P T  E X E C U T I O N  S T A R T  #################
################################################################################

# First test to make sure that wview is not running for interactive and set...
if [ -f $WVIEWD_PID -a "$1" != "get" ]; then
    echo "wviewd is running - stop wview before running this script..."
    exit 3
fi

# Make sure that the config DB is there:
if [ ! -f $WVIEW_CONF_DIR/wview-conf.sdb ]; then
    echo "wview configuration database NOT FOUND"
    exit 4
fi


METHOD=$1
SET_FILE=$2

if [ "$METHOD" = "" ]   # run interactively
then
    interactive_intro
    get_wview_conf_interactive

echo ""
echo ""
echo "################################################################################"
echo "--> wview Configuration Complete!"
echo "--> Now run wviewhtmlconfig to select your site skin."
echo "################################################################################"
else
    case "$METHOD" in
        "get" )
            print_config
            ;;
        "set" )
            if [ "$SET_FILE" = "" ]; then
                echo "set requires a source file:"
                show_usage
                exit 1
            fi
            if [ ! -f $SET_FILE ]; then
                echo "source path $SET_FILE does not exist"
                show_usage
                exit 1
            fi

            set_config_from_file
            ;;
        *)
            echo "$METHOD not supported"
            show_usage
            exit 1
            ;;
    esac
fi

exit 0

